### DISCUSSION TITLE
[csl] Issues noticed with OTBR as CSL transmitter

### DISCUSSION THREAD

**Discussion by suveshpratapa** [openthread/openthread#8141] [General]
**Description**

We're currently investigating some issues with OTBR DUT as a CSL transmitter in 1.2 low power test cases for certification.  The primary concern is inconsistency in behavior and timing calculations on OTBR.

- In certification tests, the OTBR seems to consistently miss CSL sync-times by sending too late.  Attached is a sample trace (see attached `Fail_OTS13_BR_L_5_3_1_Leader_v56.2_2022_08_13-18_12_23_Ch-11.pcapng`) where you can observe this behavior.

- I cannot reproduce the above behavior consistently on my local desk setup (ie.., non-test-harness).  The transmitter delaying too long seems to be an unintentional bug happening from time to time, and I can't seem to be able to explain it.  I can get it to happen from time to time on my desk, however the OTBR does recover.  I can force reproduction of this issue by varying log levels, which further supports the possible timing problem.

What could be causing this inconsistency on an OTBR?  Note that we perform CSL transmit calculations in software (`OT_RADIO_CAPS_TRANSMIT_TIMING=false` for our RCP).

- I'm noticing a separate issue with my reproducer on my desk.  I'm trying out regular ping packets between my OTBR and MED, and for some reason the ACK from OTBR to the ping response from the MED has `framePending=true`, which causes a data request to be sent from the MED (see attached `capture-2022_Sep_9-15_32.pcapng`).  Any idea why this could be happening?

[Fail_OTS13_BR_L_5_3_1_Leader_v56.2_2022_08_13-18_12_23_Ch-11.pcapng.zip](https://github.com/openthread/openthread/files/9538349/Fail_OTS13_BR_L_5_3_1_Leader_v56.2_2022_08_13-18_12_23_Ch-11.pcapng.zip)
[capture-2022_Sep_9-15_32.pcapng.zip](https://github.com/openthread/openthread/files/9538350/capture-2022_Sep_9-15_32.pcapng.zip)

-------------

**Reproduction steps**

- [openthread @ 2ce3d3bf0](https://github.com/openthread/openthread/commits/2ce3d3bf0)
- [ot-br-posix @ 181335224](https://github.com/openthread/ot-br-posix/commits/181335224)
- EFR32MG12 platform for both RCP and SoC

Can try test harness LP 5_3_x cases, or setup a network between OTBR and MED, with the following CSL parameters:
```
csl period 3125
csl timeout 30
```
-------------

**Comment by suveshpratapa** [comment]
We have a couple of questions about the existing implementation.
cc @Irving-cl @George-Stefan @librasungirl @edmont 

- In the sub-mac transmit code that triggers on the POSIX side, `ShouldHandleTransmitTargetTime=false` by default.  This is because the host has `RadioSupportsTransmitTiming=true`
  - `mRadioCaps=0xFF` on host because of the following implementation: [`CheckRadioCapabilities`](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel_impl.hpp#L374)->[`SPINEL_PROP_RADIO_CAPS`](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base_radio.cpp#L225)->[`otLinkRawGetCaps`->`LinkRaw::GetCaps()`](https://github.com/openthread/openthread/blob/main/src/core/mac/link_raw.hpp#L103)->[`SubMac::GetCaps()` ](https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L147) 

- Due to the above, the host calls `BeginTransmit` directly in sub-mac code. https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L442

- In our implementation, we do not have a "TxAt" implementation on our radio, so we define `OT_RADIO_CAPS_TRANSMIT_TIMING=false` for our RCP.  When the scheduled packet arrives at the RCP sub-mac, it checks RCP capabilities, therefore in our case `ShouldHandleTransmitTargetTime` becomes true for the sub-mac, and therefore it sets up delay timer in software, and then sends it out to the radio when it is time.

**Questions:**

1. If the RCP does not have `OT_RADIO_CAPS_TRANSMIT_TIMING`, like us, the delay timer happens in software on the RCP sub-mac, after which the packet is scheduled for transmission immediately.  Should the delay timer instead be set up on the host?

2. If things are working as intended and the host doesn't need to set up delay timer, then why do we bother estimating `busTxTimeUs` when calculating the `mCslFrameRequestAheadUs` for CSL transmission calculations?  https://github.com/openthread/openthread/blob/main/src/core/thread/csl_tx_scheduler.cpp#L78

**Reply by abtink** [reply]
Thanks. @suveshpratapa. I am not myself super familiar with CSL code but here is what I think/recall (which I hope helps):

> 1. If the RCP does not have OT_RADIO_CAPS_TRANSMIT_TIMING, like us, the delay timer happens in software on the RCP sub-mac, after which the packet is scheduled for transmission immediately. Should the delay timer instead be set up on the host?

I think the idea is for it to be handled on RCP. Though technically I think it can be changed to work either way. 

There is one thing to consider here: With RCP and host we may have two notions of (absolute) time (RCP time vs host time) and I guess we need to ensure the CSL timing are using the corret one (which I think is the radio/RCP time). So may make sense (and be easier) to have it be done on RCP (to avoid the need to sync and/or convert the times).

-----

> 2. If things are working as intended and the host doesn't need to set up delay timer, then why do we bother estimating busTxTimeUs when calculating the mCslFrameRequestAheadUs for CSL transmission calculations? https://github.com/openthread/openthread/blob/main/src/core/thread/csl_tx_scheduler.cpp#L78

I think this is related to interaction between `CslTxScheduler` and `Mac` layer (and then with `SubMac`). Here is what I recall:
- We have this "bus guard time" for when we ask `Mac` to start the `kOperationTransmitDataCsl` operation.
- At which point, `Mac` will ask the `CslTxScheduler` to prepare the frame.
- Then `Mac` will pass the prepared frame to `SubMac`.
- The model is that `Mac` will pass the frame to `SubMac` close to its tx time but exact tx timing is managed by `SubMac` or radio platform.
- So we need to account for bus delay in RCP model for when we start "kOperationTransmitDataCsl" operation on `Mac` to make sure we have time to pass it to RCP.

**Comment by George-Stefan** [comment]
CSL tx:
https://github.com/openthread/openthread/issues/6981#issuecomment-1040249479

This is exactly for what you see now:
**https://github.com/openthread/openthread/issues/6981#issuecomment-1040608238**


https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L428
you should get to this point with ~ OPENTHREAD_CONFIG_CSL_RECEIVE_TIME_AHEAD before CSL phase.

But this is not accurate and depends also on the CSL receiver rx margin to have a successful transmission.
In mac.cpp you have host time, in sub_mac you have RCP time.

**Reply by George-Stefan** [reply]
The 2nd issue is the Enhanced Frame Pending feature.
The only problem could be that the OTBR sets FP without a reason.
The behavior depends on https://github.com/openthread/openthread/blob/main/include/openthread/platform/radio.h#L883

**Reply by tangliligit** [reply]
from 802154-2015 csl reception: Frame Pending field in the received frame is set to one, the receiver is kept on for
macCslFramePendingWait time before going back to CSL idle listening. Otherwise, CSL idle
listening starts。Why ot send a data request？

**Reply by George-Stefan** [reply]
Read in the OT spec about CSL support.
15.4 spec is used with several Modifications

  LE specific MAC PIB attributes
• Modified: macCSLFramePendingWait SHALL
be set to '0`

**Reply by tangliligit** [reply]
get it
